caproto/_data.py

```diff
@@ -276,8 +280,12 @@ def preprocess_value(self, value):
 
         if self._max_length == 1:
             if is_array:
-                # scalar value in a list -> scalar value
-                return value[0]
+                if len(value):
+                    # scalar value in a list -> scalar value
+                    return value[0]
+                else:
+                    raise CaprotoValueError(
+                        'Cannot set a scalar to an empty array')
```

Comment:
Is this a protocol thing, an implementation detail of the c-version thing or a "This makes it better" thing?

Reply:
This also seem weird that you can init with a 0 length array, but then the first time you set it becomes a scalar.   Is it ok to have a PV that we can not re-set (from the network) back to its initial value?

Reply:
There is, admittedly, a bit of a loss of generality here. I don't personally think it's worthwhile to support the general case.

This change only affects the case where `max_length == 1` - i.e., you are working with a scalar value to start with. If we are to allow the empty array here, that would equate to passing in `None` to the putter functions that are expecting a scalar.

This is different from array values (`max_length > 1`) that can accept an array of length in `range(0, max_length + 1)`. The putter would be expected to handle arrays - including an empty array with the given `dtype`.

Reply:
I'm sold on the loss of generality, but am concerned about the interaction with https://github.com/NSLS-II/caproto/pull/405/files#diff-bc1e46712779c588b17e8e8c6c01fc23R216 which demotes (promotes) 0 length init values to scalars rather an arrays with max length of 1 (is there a difference?)

Reply:
"I'm a scalar, and my functions expect that you'll pass in a scalar" also has an implicit `min_length == 1` which is different that every other `max_length` value...

Reply:
Are you pushing for the inclusion of `min_length`, requiring our scalar-handling functions to allow `None`, or...?

Reply:
mostly just thinking (mumbling) out loud.

Reply:
I'd like to come to a consensus on these finer points before I make any additional changes...

Reply:
Maybe tracking 'scalerness' as it's own state?

Reply:
I think this is fine as is. Here is another protocol we like that has similar behavior:

```
def f(arg):
    # implicit min_length == 1
```

```
def f(*args):
    ...
```

Reply:
"you can't have an array which is 0 or 1 long because ...... reasons" is also the sort of thing that if we came across else where we would get really annoyed at.

Also related, what happens if you try to set the setpoint of a motor to a zero length array?

This code is in the path run on init (around L218 in _data) so why is this not blowing up on empy initial values?

Reply:
Let's back up a second here. When have you encountered an empty array over channel access in the wild?  If so, was it actually of any significance, uninitialized, or in error?

Reply:
Not sure I have seen one in the wild, but I think the protocol allows it...

"this PV is the list of photons I saw the last exposure" should be able to be empty? "list of files I have created from last reset"   I'm sure we could come up with some more stories with a bit of thought...

Reply:
We can revisit this at a later date. For now I'll mark this as resolved in the interest of moving forward.